chill-scala/src/main/scala/com/twitter/chill/ClosureCleaner.scala

```diff
@@ -32,233 +32,205 @@ package com.twitter.chill
+  private def getOutersOf(obj: AnyRef, hierarchy: List[(Class[_], AnyRef)] = Nil): List[(Class[_], AnyRef)] =
     outerFieldOf(obj.getClass) match {
-      case None => hierarchy // We have finished
-      case Some(f) => {
-        // f is the $outer of obj
+      case None => hierarchy
+      case Some(f) =>
         f.setAccessible(true)
-        // myOuter = obj.$outer
         val myOuter = f.get(obj)
-        // This is (Class[T], T) into the hierarchy:
-        // Note that if you use f.getType you might get an interface. No good
         val outerType = myOuter.getClass
         getOutersOf(myOuter, (outerType, myOuter) :: hierarchy)
-      }
     }
 
-  private def getInnerClassesFn(inCls: Class[_]): Set[Class[_]] = {
-    val seen = MSet[Class[_]](inCls)
-    var stack = List[Class[_]](inCls)
-    while (!stack.isEmpty) {
-      val cr = getClassReader(stack.head)
-      stack = stack.tail
-      val set = MSet[Class[_]]()
-      cr.accept(new InnerClosureFinder(set), 0)
-      for (cls <- set -- seen) {
-        seen += cls
-        stack = cls :: stack
+  private def storeAccessedFields(): Unit = {
+    if (accessedFieldsMap.isEmpty) {
+      getAccessedFields.foreach {
+        case (cls, mset) =>
+          def toF(ss: Set[String]): Set[Field] = ss.map(cls.getDeclaredField)
+          val set = mset.toSet
+          accessedFieldsMap += ((cls, toF(set)))
       }
     }
-    (seen - inCls).toSet
   }
 
-  def innerClassesOf(cls: Class[_]): Set[Class[_]] =
-    innerClasses.getOrElseUpdate(cls, getInnerClassesFn(cls))
-
-  private def getAccessedFields(cls: Class[_]): MMap[Class[_], MSet[String]] = {
-    val af = outerClassesOf(cls)
-      .foldLeft(MMap[Class[_], MSet[String]]()) { (m, clazz) =>
-        m += ((clazz, MSet[String]()))
-      }
-
-    (innerClassesOf(cls) + cls).foreach { cls =>
-      getClassReader(cls).accept(new FieldAccessFinder(af), 0)
+  private def getAccessedFields: MMap[Class[_], MSet[String]] = {
+    val af = outerClasses.foldLeft(MMap[Class[_], MSet[String]]()) { (m, cls) =>
+      m += ((cls, MSet[String]()))
     }
+    (innerClasses + funcClass).foreach(classReader(_).accept(
+      new AccessedFieldsVisitor(af)(classReader), 0))
     af
   }
 
-  /**
-   * Uses ASM to return the names of the fields accessed by this cls
-   */
-  def accessedFieldsOf(cls: Class[_]): Set[Field] = {
-    accessedFieldsMap.get(cls) match {
-      case Some(s) => s
-      case None => {
-        //Compute and store:
-        val af = getAccessedFields(cls)
-        def toF(ss: Set[String]): Set[Field] = ss.map { cls.getDeclaredField(_) }
-
-        val s = af.get(cls).map { _.toSet }.getOrElse(Set[String]())
-        // Add all of af:
-        af.foreach { clsMSet =>
-          val set = clsMSet._2.toSet
-          accessedFieldsMap += ((clsMSet._1, toF(set)))
-        }
-        toF(s)
+  private def innerClasses: Set[Class[_]] = {
+    val seen = MSet[Class[_]](funcClass)
+    var stack = MStack[Class[_]](funcClass)
```

Comment:
we don't need a `var` to a mutable stack do we? I think a `val` here is fine and no need to reassign below (or use a `var` to a `List[Class[_]]` as the stack, or use a tail-recursive loop and no var or mutable. :)